rhttp/client.go

```diff
@@ -0,0 +1,233 @@
+package rhttp
+
+import (
+	"io"
+	"math"
+	"math/rand"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// Client is an extension of the net/http client with ability to add
+// timeouts and exponential backoff
+type Client struct {
+	MaxRetries         int
+	ExponentialBackoff bool
+	RetryTime          time.Duration
+
+	HTTPClient *http.Client
+}
+
+// DefaultClient is a go-ns specific http client with sensible timeouts
+// and exponential backoff
+var DefaultClient = &Client{
+	MaxRetries:         10,
+	ExponentialBackoff: true,
+	RetryTime:          20 * time.Millisecond,
+
+	HTTPClient: &http.Client{
+		Timeout: 10 * time.Second,
+		Transport: &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout: 5 * time.Second,
+			}).Dial,
+			TLSHandshakeTimeout: 5 * time.Second,
+			MaxIdleConns:        10,
+			IdleConnTimeout:     30 * time.Second,
+		},
+	},
+}
+
+// Do calls net/http Do with the addition of exponential backoff
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
+	resp, err := c.HTTPClient.Do(req)
+	if err != nil {
+		if c.ExponentialBackoff {
+			return c.doWithBackoff(req)
+		}
+		return nil, err
+	}
+
+	if resp.StatusCode >= http.StatusInternalServerError && c.ExponentialBackoff {
+		return c.doWithBackoff(req)
+	}
+
+	return resp, err
+}
+
+func (c *Client) doWithBackoff(req *http.Request) (resp *http.Response, err error) {
+	for i := 0; i < c.MaxRetries; i++ {
```

Comment:
if this was `i := 1; i <= c.MaxRetries` you wouldn't need to create the attempt var